persistence-api/src/main/java/io/stargate/db/query/BoundSelect.java

```diff
@@ -0,0 +1,39 @@
+package io.stargate.db.query;
+
+import io.stargate.db.schema.Column;
+import io.stargate.db.schema.Table;
+import java.util.OptionalLong;
+import java.util.Set;
+
+/**
+ * A fully bound SELECT query.
+ *
+ * <p>Please note that the method of this class currently only expose a <b>very</b> restricted
+ * subset of what a SELECT query may contain. This is meant to evolve over time so more details are
+ * exposed.
+ */
+public interface BoundSelect extends BoundQuery {
+  Table table();
+
+  /**
+   * The subset of the {@link #table} columns that are queried by this SELECT, or an empty set for
+   * "star" SELECT.
+   *
+   * <p>Please note that this does not necessarily directly map to the entries in the result set of
+   * this SELECT, as the result set may have entries that are not just a table column. For instance,
+   * for {@code SELECT x, SUM(y + x), TTL(z) FROM ...}, this method will return a set containing
+   * columns x, y and z (but the result sets are more "composite" entries).
+   *
+   * @return a set with the columns of {@link #table} that this SELECT has to query, or an empty set
+   *     if this is a "star" select ({@code SELECT * FROM ...}).
+   */
+  Set<Column> selectedColumns();
```

Comment:
I suppose the set of selected columns is available in built select queries too. Why are we only exposing it in the "bound" statement but not in the "built" statement? I imagine there may be use cases for examining selected columns before binding statement parameters.

Reply:
To be more specific, I'm wondering about the difference between "built" statements where we use concrete classes to the various statement sub-types, and the "bound" statements where we use interfaces for statements sub-types.

Reply:
So first, technically, `BuiltSelect` does expose the selected columns (`BuiltSelect#selectColumns` exists and is public). So code can examine that if it so wishes.

But to answer the more general question: the focus of this patch is to start introduce ways to manipulate queries within the `DataStore`, to have a way to have generic "middle-tier" logic that sit between the top-level APIs and the low-level persistence. That's `BoundQuery` and that's why I focused on exposing things here and provide clear public interfaces.

In contrast, the query builder is just one tool to build queries. Eventually, we'll also be able to create query objects through parsing (and that may or may not use the query builder underneath at all). So I bothered less about interfaces and exposing stuffs on "built" queries (but again, selected columns _are_ exposed).

More pragmatically, I expect we "may" have multiple implementations of the various `BoundQuery` (or maybe not, maybe query parsing and any other needs will also use the query builder, I don't know yet), but I don't really expect we'll bother with having multiple different query builder implementation (hence the interface on one side, concrete classes on the other).

Reply:
Thanks for the explanation, @pcmanus !

I was approaching this from the SQL perspective, where I thought it might be beneficial to have more concrete query types returned from the builder to avoid casting to `BuiltSelect` (e.g. `builder.select()...buid()` would return something like a `SelectQuery` as opposed to generic `Query`). Yet, I do not have a specific use case at this point, so current state of the code LGTM.